### PR TITLE
Initial work for introducing openssl/boringssl as a TLS provider

### PR DESCRIPTION
**Pull request by arnarpall**
In order be able to support running linkerd in fips compliant mode we need replace rustls and ring with openssl
since *ring* is not fips validated/compliant.

* Added support for openssl
* Testing out Boringssl
* Compilation features flags (openssl-tls, rustls-tls, boring-tls)

### CODE REVIEW

File: linkerd/tls/src/imp/boring.rs
Comment (arnarpall) [inline linkerd/tls/src/imp/boring.rs:39]: This needs some cleaning up

File: linkerd/tls/src/imp/boring.rs
Comment (arnarpall) [inline linkerd/tls/src/imp/boring.rs:110]: Same as above, error handling cleanup.

File: linkerd/tls/src/imp/boring.rs
Comment (arnarpall) [reply linkerd/tls/src/imp/boring.rs]: Error handling in the boring crate is being revamped here: https://github.com/cloudflare/boring/pull/23

Comment (nox) [reply linkerd/tls/src/imp/boring.rs]: I'll revisit that PR at a later time because I have some other stuff on my plate and I misunderstood some critical parts of AsyncRead hah.

File: Dockerfile
Comment (olix0r) [inline Dockerfile:43]: Do we really need golang here?

File: .gitignore
Comment (olix0r) [inline .gitignore:7]: generally prefer to keep these in your personal ~/.gitignore. Not a huge deal, but in interest of keeping the PR small and focused...
```suggestion
```

File: Cargo.lock
Comment (olix0r) [inline Cargo.lock:299]: We shouldn't nee to pull in env-logger... Is this coming from one of the new dependencies?

env-logger should only be needed by binary entry points and not by libraries. And for libraries, the `tracing` crate is definitely preferable.

File: Cargo.lock
Comment (olix0r) [inline Cargo.lock:26]: Similarly, this pulls in a second (older) version of `ansi_term`. It would be best if we could avoid these duplicate deps.

File: Cargo.lock
Comment (olix0r) [inline Cargo.lock:213]: This is a command-line parsing crate. Why is this getting pulled in?

File: Cargo.lock
Comment (olix0r) [inline Cargo.lock:158]: It looks like boring-sys should use bindgen without default features (https://github.com/rust-lang/rust-bindgen/blob/b60339ece3994868a55889b7f1f6043ab29ba30e/Cargo.toml#L74-L80) to avoid pulling in unnecessary dependencies.

**Review by olix0r — CHANGES_REQUESTED** [review]
I'll continue looking at the feature flagging scheme to see what changes we can make to our APIs to support this better. In the meantime it would be great to minimize the dependencies for boring-sys.

**Comment by olix0r** [comment]
Also note that we'll need some DCO sign-off for this change. See https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md#developer-certificate-of-origin. Note that it's not necessary to squash/amend all commits with the sign-off if long as you make a public statement in the PR.

File: Cargo.lock
Comment (olix0r) [reply Cargo.lock:158]: Specifically, pulling in `log` is okay; but pulling in `env_logger` is undesirable. It seems, though, that bindgen doesn't expose distinct dependencies.

Is bindgen actually a runtime dependency or is it only a build-time dependency? If the latter, it should probably be in `dev-dependencies` and not `dependencies`.

File: Dockerfile
Comment (arnarpall) [reply Dockerfile:43]: Ah yeah that's crazy, the boring-sys crate uses go as part of the whole build process.

We probably have to figure out a better way to handle this docker file.

File: .gitignore
Comment (arnarpall) [reply .gitignore:7]: Right good point.

**Comment by arnarpall** [comment]
I agree to the DCO for all the commits in this PR.

**Comment by olix0r** [comment]
It looks like the `linkerd/identity` crate's dependencies need some fixing:

```
:; (cd linkerd/identity && cargo check)
...
error: aborting due to 25 previous errors

Some errors have detailed explanations: E0412, E0433.
For more information about an error, try `rustc --explain E0412`.
error: could not compile `linkerd-identity`
```

I suspect related to feature-flagging and conditional compilation.

File: Cargo.lock
Comment (olix0r) [reply Cargo.lock:158]: Ah yeah, I see now that it is indeed a build-only dependency in boring-sys. It's probably less of an issue if these aren't actually pulled into the runtime, then.

File: linkerd/identity/src/imp/openssl.rs
Comment (hawkw) [inline linkerd/identity/src/imp/openssl.rs:180]: nit: again, I think a lot of this boilerplate could be generated using `thiserror`, like so
```suggestion
#[derive(Clone, Debug. Error)]
pub enum InvalidCrt {
    #[error("subject alt name incorrect ({0})"]
    SubjectAltName(String),
    #[error("{0}")]
    Verify(#[source] X509VerifyResult),
    #[error(transparent)]
    General(#[from] Error),
}

impl From<ErrorStack> for InvalidCrt {
    fn from(err: ErrorStack) -> Self {
        InvalidCrt::General(err.into())
    }
}
```

File: linkerd/identity/src/imp/openssl.rs
Comment (hawkw) [inline linkerd/identity/src/imp/openssl.rs:227]: again, if the intention is to abstract over all the openssl-specific implementation details, we should ideally find a way to make these not be `pub`...

File: linkerd/identity/src/imp/openssl.rs
Comment (hawkw) [inline linkerd/identity/src/imp/openssl.rs:293]: nit, take it or leave it: it might be slightly more idiomatic if this was an implementation of the `Default` trait? but, `empty()` is also fine...

File: linkerd/identity/src/imp/openssl.rs
Comment (hawkw) [inline linkerd/identity/src/imp/openssl.rs:269]: it might be a little more idiomatic to make this
```suggestion
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        f.debug_struct("ClientConfig").field("protocols", &self.protocols).finish()
    }
```
so that it's clearer which type this is?

File: linkerd/identity/src/imp/openssl.rs
Comment (hawkw) [inline linkerd/identity/src/imp/openssl.rs:343]: nit: prefer
```suggestion

        f.debug_struct("ServerConfig")
            .field("alpn_protocols", &self.alpn_protocols)
            .field("key", &self.key)
            .finish()
```

File: linkerd/identity/src/imp/rustls.rs
Comment (hawkw) [inline linkerd/identity/src/imp/rustls.rs:157]: currently, the rustls version of this type does not include the inner value:
https://github.com/linkerd/linkerd2-proxy/blob/c188444e805787a6e271782bba2442f2a8ed0387/linkerd/identity/src/lib.rs#L306-L310

what's the rationale behind changing this code? do we need to print this for debugging?

File: linkerd/identity/src/imp/rustls.rs
Comment (hawkw) [inline linkerd/identity/src/imp/rustls.rs:175]: nit: could use `thiserror` here:
```suggestion
#[derive(Clone, Debug, Error)]
#[error(transparent)]
pub struct InvalidCrt(#[from] rustls::TLSError);
```

File: linkerd/identity/src/imp/rustls.rs
Comment (hawkw) [inline linkerd/identity/src/imp/rustls.rs:368]: i think this should be
```suggestion
        f.debug_struct("ClientConfig")
```

File: linkerd/identity/src/imp/rustls.rs
Comment (hawkw) [inline linkerd/identity/src/imp/rustls.rs:410]: and this should be 
```suggestion
        f.debug_struct("ServerConfig")
```

File: linkerd/identity/src/lib.rs
Comment (hawkw) [inline linkerd/identity/src/lib.rs:52]: these impls could also be generated using `thiserror`

```rust
/// An error returned from the TLS implementation.
#[derive(Error)]
#[error(transparent)]
pub struct Error(#[from] imp::Error);
```

File: linkerd/identity/src/lib.rs
Comment (hawkw) [inline linkerd/identity/src/lib.rs:71]: nit: the `From` impl could be derived by `thiserror`:
```suggestion
pub struct InvalidCrt(#[from] imp::InvalidCrt);

impl From<imp::InvalidCrt> for InvalidCrt {
    fn from(err: imp::InvalidCrt) -> Self {
        InvalidCrt(err)
    }
}
```

File: linkerd/identity/src/lib.rs
Comment (hawkw) [inline linkerd/identity/src/lib.rs:310]: let's remove the commented-out code?

File: linkerd/identity/src/lib.rs
Comment (hawkw) [inline linkerd/identity/src/lib.rs:319]: nit: prefer whitespace here
```suggestion
    }
    
    pub fn set_protocols(&mut self, protocols: Vec<Vec<u8>>) {
```

File: linkerd/identity/src/imp/openssl.rs
Comment (hawkw) [inline linkerd/identity/src/imp/openssl.rs:273]: it looks like the `new` constructor is never called outside this module, can we just change this to
```suggestion
    fn new(
```
since this constructor exposes an implementation detail (the `X509Store` type from `openssl`).

File: linkerd/identity/src/imp/openssl.rs
Comment (hawkw) [inline linkerd/identity/src/imp/openssl.rs:263]: these are never accessed outside this module, they shouldn't need to be `pub`:
```suggestion
    root_certs: Arc<X509Store>,
    key: Option<Arc<Key>>,
    cert: Option<Arc<Crt>>,
    protocols: Arc<Vec<Vec<u8>>>,
```

File: linkerd/identity/src/imp/openssl.rs
Comment (hawkw) [inline linkerd/identity/src/imp/openssl.rs:309]: this constructor isn't called outside this module (afaict), it shouldn't need to be `pub`:
```suggestion
    fn new(
```

File: linkerd/identity/src/imp/openssl.rs
Comment (hawkw) [inline linkerd/identity/src/imp/openssl.rs:305]: these shouldn't need `pub`
```suggestion
    root_certs: Arc<X509Store>,
    key: Option<Arc<Key>>,
    cert: Option<Arc<Crt>>,
    alpn_protocols: Arc<Vec<Vec<u8>>>,
```

File: linkerd/tls/src/imp/openssl.rs
Comment (hawkw) [inline linkerd/tls/src/imp/openssl.rs:49]: in what cases will these `unwrap`s fail? if they should never fail in normal operation, these should probably be `expect`s with messages explaining why; otherwise, we should return an error in this case.

File: linkerd/tls/src/imp/openssl.rs
Comment (hawkw) [inline linkerd/tls/src/imp/openssl.rs:66]: what is this code doing? what does `err.ssl()` do, and why do we ignore the return value?

File: linkerd/tls/src/imp/openssl.rs
Comment (hawkw) [inline linkerd/tls/src/imp/openssl.rs:86]: let's either remove the `println!` or replace it with a `debug!` or something, please :)

File: linkerd/tls/src/imp/openssl.rs
Comment (hawkw) [inline linkerd/tls/src/imp/openssl.rs:104]: we should probably remove the `println!`s here as well (or replace them with logging)

File: linkerd/tls/src/imp/openssl.rs
Comment (hawkw) [inline linkerd/tls/src/imp/openssl.rs:117]: ...and here

File: linkerd/tls/src/imp/openssl.rs
Comment (hawkw) [inline linkerd/tls/src/imp/openssl.rs:183]: let's remove the `println!`s or replace them with `debug!` or `trace!` logging?

File: linkerd/tls/src/imp/openssl.rs
Comment (hawkw) [inline linkerd/tls/src/imp/openssl.rs:165]: let's remove the `println!` or replace it with logging here?

File: linkerd/tls/src/imp/openssl.rs
Comment (hawkw) [inline linkerd/tls/src/imp/openssl.rs:194]: another `println!`

File: linkerd/tls/src/imp/openssl.rs
Comment (hawkw) [inline linkerd/tls/src/imp/openssl.rs:229]: nit: this `match` could be simplified to
```suggestion
        ClientId::from_str(peer.dnsname().unwrap()).ok()
```

**Review by hawkw — COMMENTED** [review]
i gave this a quick review focusing mostly on style and idiomatic Rust --- I'd like to look closer at the feature-flagging strategy here in the future. the overall strategy --- incapsulating `rustls`/`openssl`/`boringssl`-specific implementation details in private modules that are wrapped by backend-agnostic types to expose them to the rest of the codebase --- definitely looks correct!

some general notes:

- i noticed there's a very large number of `unwrap()`s in this PR. `unwrap` will cause the process to panic (crash instantly). 

  in general, we should try to avoid using `unwrap()`, even for errors that are unrecoverable and _should_ cause the proxy to panic --- we should prefer `expect` for fatal errors, because `expect` includes a message. in general, we should try to include messages for these errors explaining why the error is fatal and/or why we expect it to not occur in normal operation. 

- also, i noticed that there are a lot of `println!`s, presumably for locally debugging while working on this change. before we merge this, we should try to remove those, or replace them with logging if that information is valuable to record in production.

- some error-handling boilerplate could probably be simplified a lot by using the [`?` operator](https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/the-question-mark-operator-for-easier-error-handling.html) in more places

File: linkerd/identity/Cargo.toml
Comment (arnarpall) [inline linkerd/identity/Cargo.toml:17]: I added the `features = ["std"]` in order to satisfy `thiserror::Error` not sure if we want this around but I'll circle back to it.

File: linkerd/identity/src/imp/openssl.rs
Comment (arnarpall) [reply linkerd/identity/src/imp/openssl.rs]: Yeah this was not really intentional but rather just inexperience. I've applied your suggestions 👍

File: linkerd/identity/Cargo.toml
Comment (hawkw) [reply linkerd/identity/Cargo.toml:17]: this is fine; linkerd is always built with the Rust standard library, so enabling `ring`'s dependency on `std` is fine.

**Comment by arnarpall** [comment]
I just want to thank you for your review @hawkw it's immensely helpful.

I've been working my way through your comments and hopefully I should be able to get something together today.

* Usages of `println` have been chanaged to  either `debug` or `trace`. This was really just an oversight on my part.
* `Error` trait boilerplate removed and will now be generated by `thiserror:Error`

File: linkerd/identity/src/imp/openssl.rs
Comment (arnarpall) [reply linkerd/identity/src/imp/openssl.rs:111]: Ah yeah thanks for this. I really really disliked this block of code when I wrote it.

File: linkerd/identity/src/imp/openssl.rs
Comment (arnarpall) [reply linkerd/identity/src/imp/openssl.rs:191]: Yeah I'm going to circle back to this.
The main thing here is the coupling between the Identity and the Tls modules respectively.
Ideally these should be completely hidden from the consumer. I'll figure something out in this regard :D

**Comment by hawkw** [comment]
> I just want to thank you for your review @hawkw it's immensely helpful.

glad to help --- let me know (or click the "re-request review" button) whenever it's ready for another look?

File: linkerd/identity/src/imp/openssl.rs
Comment (arnarpall) [reply linkerd/identity/src/imp/openssl.rs:33]: Yeah I totally agree.
I'm looking into what is the best solution here. The reason why this is public at this time is because of the identity vs tls modules.

When constructing a TlsConnector I need access to the primitive type Pkey<Private> from within the tls module.
I'm pretty sure this is something that can be solved by implementing `From` and `Into` traits but I'm not seeing the solution at this time.

File: linkerd/identity/src/imp/openssl.rs
Comment (arnarpall) [reply linkerd/identity/src/imp/openssl.rs:239]: Absolutely.

File: linkerd/identity/src/imp/rustls.rs
Comment (arnarpall) [reply linkerd/identity/src/imp/rustls.rs:136]: Very good point. I added this when I was debugging something myself and forgot to remove it.

File: linkerd/tls/src/imp/openssl.rs
Comment (arnarpall) [reply linkerd/tls/src/imp/openssl.rs:105]: I'll change this tomorrow morning. I was battling a little bit with the returned error from tokio_boring

**Comment by arnarpall** [comment]
@hawkw @olix0r  could you guys give the pull request another look over ?

I believe I've fixed most of your concerns @hawkw aside from the public fields in a few places. It would be awesome if we could perhaps try to sync up at some point and figure out a good way to hide these fields.

I have not fixed the build/dependency issues that you mentioned @olix0r I will circle back to that.

**Comment by olix0r** [comment]
Last week I spent a few days trying to rework the TLS and identity modules. That work is incomplete but shared [here](https://github.com/linkerd/linkerd2-proxy/pull/new/ver/tls-generis). The summary is;

* The TLS module is generic over the actual TLS implementation (which holds the config types -- these should not be exposed generally).
* The identity module holds only the core structs & traits needed to implement identity. These traits model clients and servers for TLS initiation and termination.
* the proxy-identity module communicates with the API but delegates validation to the underlying identity implementation.
* at startup, the proxy is configured with an identity mode that determines which implementation is used from the implementations compiled with the proxy. This lets us play well with the Rust feature system so that features are purely additive and do not conflict.

I had to put this down to focus on inbound policy changes, but I should be able to revive this work once our access policy work lands.

**Comment by olix0r** [comment]
Closing this out in favor of #1351, which takes advantage of recent API changes to support compilation with `--all-features` (i.e. `rustls` and `boring` are not mutually exclusive).

There are some issues we'll need to resolve before merging that. Notably: ALPN support & trust anchor parsing

**Pull request closed without merging**
